### ISSUE TITLE
Creating files with reserved Windows file names creates issues within Windows

### ISSUE THREAD

**Issue by Dr-Irv** [Microsoft/WSL#2689]
* Your Windows build number:  Microsoft Windows [Version 10.0.15063]

* What you're doing and what's happening:  
`echo "foobar" > con.out`

This creates the file con.out in the current directory.  It is totally usable from bash.  

But from DOS, you cannot look at the file, rename the file, or do anything with it, because "CON" is a reserved file name.  You cannot even rename the file from Windows Explorer.

* What's wrong / what should be happening instead:
Maybe a warning should be issued if from Linux, you are creating a file that has a name that is invalid for Windows.  Or it should not be possible to create files with the reserved names.

Apparently, these are the reserved file names:

CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, and LPT9  (see https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx)

**Comment by therealkenc** [comment]
This is spiritually similar to #2081. See the notes on namespaces pasted from your link below. Those names are reserved in Win32 by convention, but not the NTFS filesystem.

You _could_ fail on the names (and similarly fail on case collisions), controlled with a DrvFS mount option as proposed [here](https://github.com/Microsoft/WSL/issues/2081#issuecomment-302569755). [An option that would be off by default to avoid "why can't I `echo foobar > con.out` on DrvFs?" bug reports. And turned on by no one. But I digress.]
<br>
> Namespaces

> There are two main categories of namespace conventions used in the Windows APIs, commonly referred to as NT namespaces and the Win32 namespaces. The NT namespace was designed to be the lowest level namespace on which other subsystems and namespaces could exist, including the Win32 subsystem and, by extension, the Win32 namespaces. POSIX is another example of a subsystem in Windows that is built on top of the NT namespace. Early versions of Windows also defined several predefined, or reserved, names for certain special devices such as communications (serial and parallel) ports and the default display console as part of what is now called the NT device namespace, and are still supported in current versions of Windows for backward compatibility.

**Comment by therealkenc** [comment]
This is spiritually similar to #2081. The names are reserved by convention in Win32, but are legal NTFS filenames. See comments on namespaces from your link at the bottom.

Your ask could be provided with a DrvFs mount option (similar to disallowing case collisions) per the proposal [here](https://github.com/Microsoft/WSL/issues/2081#issuecomment-302569755).

<br>

>Namespaces

>There are two main categories of namespace conventions used in the Windows APIs, commonly referred to as NT namespaces and the Win32 namespaces. The NT namespace was designed to be the lowest level namespace on which other subsystems and namespaces could exist, including the Win32 subsystem and, by extension, the Win32 namespaces. POSIX is another example of a subsystem in Windows that is built on top of the NT namespace. Early versions of Windows also defined several predefined, or reserved, names for certain special devices such as communications (serial and parallel) ports and the default display console as part of what is now called the NT device namespace, and are still supported in current versions of Windows for backward compatibility.

**Comment by Vaporeon** [comment]
Can I request this **NOT** be "fixed".

I have had a problem with software that happens to be more that happy to create files with invalid names.

CMD can't touch them
Powershell can't touch them
Win32 explorer can't touch them
Trying to interact with the file using the 8:3 compatibility name does not work
Running chkdsk shows no errors
Deleting the directory one level up leaves the files behind

The files are effectively stuck and it seems there was no way to remove or rename it other than booting into Linux, mounting the drive and manipulating the file that way. Now however I can just start a bash shell and mv "invalidname.foo" win32safename.foo and everything is fine. It has been really convenient.

**Comment by therealkenc** [comment]
> Can I request this **NOT** be "fixed". [...] It has been really convenient.

Agreed, but your line reasoning failed last time I tried to evoke the same. In ntfs-3g the mount option is called [`windows_names`](http://manpages.ubuntu.com/manpages/zesty/man8/ntfs-3g.8.html). 

So, the solution seems to be to make it an obscure mount option, make posix (ie current) behaviour the default, and then never use it.

```
   Windows Filename Compatibility
       NTFS  supports several filename namespaces: DOS, Win32 and POSIX. While
       the ntfs-3g driver handles all of them, it always creates new files  in
       the  POSIX  namespace  for  maximum  portability  and  interoperability
       reasons.   This  means  that  filenames  are  case  sensitive  and  all
       characters  are allowed except '/' and '\0'. This is perfectly legal on
       Windows,  though  some  application  may  get  confused.   The   option
       windows_names  may  be  used  to apply Windows restrictions to new file
       names.

       windows_names
              This option prevents files, directories and extended  attributes
              to be created with a name not allowed by windows, because

                     - it contains some not allowed character,
                     - or the last character is a space or a dot,
                     - or the name is reserved.

              The forbidden characters are the nine characters " * / : < > ? \
              | and those whose code is less than 0x20, and the reserved names
              are  CON,  PRN, AUX, NUL, COM1..COM9, LPT1..LPT9, with no suffix
              or followed by a dot.

              Existing such files can still be read (and renamed).
```

**Comment by poizan42** [comment]
@Vaporeon
https://gist.github.com/poizan42/9ef1cdfb16e2515b586719900072599d
https://gist.github.com/poizan42/ed45fae25e6c482a3d9fe2bdac84894b
